executor/write.go

```diff
@@ -933,43 +977,183 @@ func getKeysNeedCheck(ctx sessionctx.Context, t table.Table, rows [][]types.Datu
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
 
 	// batch get values
 	nKeys := 0
-	for _, v := range rowWithKeys {
+	for _, v := range keysInRows {
 		nKeys += len(v)
 	}
 	batchKeys := make([]kv.Key, 0, nKeys)
-	for _, v := range rowWithKeys {
+	for _, v := range keysInRows {
 		for _, k := range v {
 			batchKeys = append(batchKeys, k.key)
 		}
 	}
 	values, err := ctx.Txn().GetSnapshot().BatchGet(batchKeys)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	return keysInRows, values, nil
+}
+
+func (e *InsertExec) checkBatchLimit() error {
+	sessVars := e.ctx.GetSessionVars()
+	batchInsert := sessVars.BatchInsert && !sessVars.InTxn()
+	batchSize := sessVars.DMLBatchSize
+	if batchInsert && e.rowCount >= batchSize {
+		e.ctx.StmtCommit()
+		if err := e.ctx.NewTxn(); err != nil {
+			// We should return a special error for batch insert.
+			return ErrBatchInsertFail.Gen("BatchInsert failed with error: %v", err)
+		}
+		e.rowCount = 0
+		if !sessVars.ImportingData {
+			sessVars.GetWriteStmtBufs().BufStore = kv.NewBufferStore(e.ctx.Txn(), kv.TempTxnMemBufCap)
+		}
+	}
+	return nil
+}
+
+func (e *InsertExec) initDirtyRowValue(newRows [][]types.Datum, keysInRows [][]keyWithDupError,
+	values map[string][]byte) (err error) {
+	e.dirtyRowValue = make(map[string][]byte, len(newRows))
+	handles := make([]int64, 0, len(newRows))
+	for _, keysInRow := range keysInRows {
+		for _, k := range keysInRow {
+			if val, found := values[string(k.key)]; found {
+				var handle int64
+				if k.isRecordKey {
+					handle, err = tablecodec.DecodeRowKey(k.key)
+					if err != nil {
+						return errors.Trace(err)
+					}
+				} else {
+					handle, err = tables.DecodeHandle(val)
+					if err != nil {
+						return errors.Trace(err)
+					}
+				}
+				handles = append(handles, handle)
+			}
+		}
+	}
+	e.dirtyRowValue, err = getOldValues(e.ctx, e.Table, handles)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (e *InsertExec) batchUpdateDupRows(newRows [][]types.Datum, onDuplicate []*expression.Assignment) ([][]types.Datum, error) {
+	keysInRows, values, err := batchGetInsertRows(e.ctx, e.Table, newRows)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
+	err = e.initDirtyRowValue(newRows, keysInRows, values)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for i, keysInRow := range keysInRows {
+		for _, k := range keysInRow {
+			if val, found := values[string(k.key)]; found {
+				var handle int64
+				if k.isRecordKey {
+					handle, err = tablecodec.DecodeRowKey(k.key)
+					if err != nil {
+						return nil, errors.Trace(err)
+					}
+				} else {
+					handle, err = tables.DecodeHandle(val)
+					if err != nil {
+						return nil, errors.Trace(err)
+					}
+				}
+				oldValue, ok := e.dirtyRowValue[string(e.Table.RecordKey(handle))]
+				if !ok {
+					return nil, errors.NotFoundf("can not be duplicated row, due to old row not found. handle %d", handle)
```

Comment:
`dirtyRowValue` is map prepared in `initDirtyRowValue`, right? If you cannot find such key in map, it means some fatal mistake happens. Error message should be more aggressive.

Reply:
emmm, I think the message is suitable. Or any suggestions?

Reply:
maybe `errors.Fataf` ?